[CI] ILMDownsampleDisruptionIT testILMDownsampleRollingRestart failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #11230 / almalinux-8-aarch64_checkpart2_platform-support-arm](https://gradle-enterprise.elastic.co/s/orvciwoqmi2ma)
- [elasticsearch-pull-request #99182 / part-2](https://gradle-enterprise.elastic.co/s/pd73hytrz34yg)
- [elasticsearch-pull-request #99113 / part-2](https://gradle-enterprise.elastic.co/s/sxk3jv5vdwz3e)
- [elasticsearch-pull-request #99003 / part-2](https://gradle-enterprise.elastic.co/s/kuend2slyn5le)
- [elasticsearch-pull-request #98995 / part-2](https://gradle-enterprise.elastic.co/s/ejfpvf6cwp3v4)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:downsample:internalClusterTest" --tests "org.elasticsearch.xpack.downsample.ILMDownsampleDisruptionIT.testILMDownsampleRollingRestart" -Dtests.seed=30E0CBFB781FCAF4 -Dtests.locale=es-ES -Dtests.timezone=Etc/GMT-5 -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.downsample.ILMDownsampleDisruptionIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testILMDownsampleRollingRestart'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: Shard [downsample-1h-xncwqbcsbt][0] is still locked after 5 sec waiting
```

**Issue Reasons:**
- [main] 5 failures in test testILMDownsampleRollingRestart (2.2% fail rate in 230 executions)
- [main] 4 failures in step part-2 (3.3% fail rate in 120 executions)
- [main] 4 failures in pipeline elasticsearch-pull-request (3.3% fail rate in 120 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -510,9 +510,6 @@ tests:
 - class: org.elasticsearch.xpack.security.CoreWithSecurityClientYamlTestSuiteIT
   method: test {yaml=indices.validate_query/20_query_string/validate_query with query_string parameters}
   issue: https://github.com/elastic/elasticsearch/issues/137391
-- class: org.elasticsearch.xpack.downsample.ILMDownsampleDisruptionIT
-  method: testILMDownsampleRollingRestart
-  issue: https://github.com/elastic/elasticsearch/issues/136585
 
 # Examples:
 #

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -14,7 +14,6 @@ import org.elasticsearch.action.admin.indices.get.GetIndexResponse;
 import org.elasticsearch.action.admin.indices.settings.get.GetSettingsRequest;
 import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest;
 import org.elasticsearch.action.downsample.DownsampleConfig;
-import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.core.TimeValue;
@@ -52,7 +51,6 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import static org.elasticsearch.core.Strings.format;
@@ -61,7 +59,6 @@ import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertResp
 import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
 import static org.elasticsearch.xpack.core.rollup.ConfigTestHelpers.randomInterval;
 import static org.hamcrest.Matchers.equalTo;
-import static org.hamcrest.Matchers.notNullValue;
 
 @ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, numClientNodes = 4)
 public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -176,8 +176,7 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         // We assert that ILM successfully completed the phase
         logger.info("Waiting for ILM to complete the phase for index [{}]", targetIndex);
         ClusterServiceUtils.addTemporaryStateListener(clusterState -> {
-            IndexMetadata indexMetadata = getIndexMetadataByIndexName(clusterState, targetIndex);
-            // We expect a single project available in this cluster
+            IndexMetadata indexMetadata = clusterState.metadata().getProject().index(targetIndex);
             return indexMetadata.getLifecycleExecutionState() != null
                 && Objects.equals(indexMetadata.getLifecycleExecutionState().step(), PhaseCompleteStep.NAME);
         });
@@ -228,20 +227,5 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
                 assertTrue(targetIndexSearch.getHits().getHits().length > 0);
             }
         );
-    }
-
-    /**
-     * We assume the index exists and its index name is unique.
-     * @return the index metadata found in any project that contains an index with this name.
-     */
-    private IndexMetadata getIndexMetadataByIndexName(ClusterState clusterState, String indexName) {
-        AtomicReference<IndexMetadata> indexMetadata = new AtomicReference<>();
-        clusterState.forEachProject(projectState -> {
-            if (projectState.metadata().hasIndex(indexName)) {
-                indexMetadata.set(projectState.metadata().index(indexName));
-            }
-        });
-        assertThat(indexMetadata.get(), notNullValue());
-        return indexMetadata.get();
     }
 }

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -169,7 +169,7 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         startDownsampleTaskViaIlm(sourceIndex, targetIndex);
         assertBusy(() -> assertTargetIndex(cluster, targetIndex, indexedDocs));
         ensureGreen(targetIndex);
-        // We assert that ILM successfully completed the phase
+        // We wait for ILM to successfully complete the phase
         logger.info("Waiting for ILM to complete the phase for index [{}]", targetIndex);
         awaitClusterState(clusterState -> {
             IndexMetadata indexMetadata = clusterState.metadata().getProject().index(targetIndex);

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -23,7 +23,6 @@ import org.elasticsearch.index.mapper.DateFieldMapper;
 import org.elasticsearch.index.query.MatchAllQueryBuilder;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
-import org.elasticsearch.test.ClusterServiceUtils;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.InternalTestCluster;
 import org.elasticsearch.xcontent.XContentBuilder;
@@ -172,7 +171,7 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         ensureGreen(targetIndex);
         // We assert that ILM successfully completed the phase
         logger.info("Waiting for ILM to complete the phase for index [{}]", targetIndex);
-        ClusterServiceUtils.addTemporaryStateListener(clusterState -> {
+        awaitClusterState(clusterState -> {
             IndexMetadata indexMetadata = clusterState.metadata().getProject().index(targetIndex);
             return indexMetadata.getLifecycleExecutionState() != null
                 && Objects.equals(indexMetadata.getLifecycleExecutionState().step(), PhaseCompleteStep.NAME);

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -25,15 +25,19 @@ import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.InternalTestCluster;
-import org.elasticsearch.test.junit.annotations.TestIssueLogging;
 import org.elasticsearch.xcontent.XContentBuilder;
 import org.elasticsearch.xcontent.XContentFactory;
 import org.elasticsearch.xpack.aggregatemetric.AggregateMetricMapperPlugin;
 import org.elasticsearch.xpack.ccr.Ccr;
 import org.elasticsearch.xpack.core.LocalStateCompositeXPackPlugin;
+import org.elasticsearch.xpack.core.ilm.ExplainLifecycleRequest;
+import org.elasticsearch.xpack.core.ilm.ExplainLifecycleResponse;
+import org.elasticsearch.xpack.core.ilm.IndexLifecycleExplainResponse;
 import org.elasticsearch.xpack.core.ilm.LifecyclePolicy;
 import org.elasticsearch.xpack.core.ilm.LifecycleSettings;
 import org.elasticsearch.xpack.core.ilm.Phase;
+import org.elasticsearch.xpack.core.ilm.PhaseCompleteStep;
+import org.elasticsearch.xpack.core.ilm.action.ExplainLifecycleAction;
 import org.elasticsearch.xpack.core.ilm.action.ILMActions;
 import org.elasticsearch.xpack.core.ilm.action.PutLifecycleRequest;
 import org.elasticsearch.xpack.ilm.IndexLifecycle;
@@ -57,6 +61,7 @@ import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertResp
 import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
 import static org.elasticsearch.xpack.core.rollup.ConfigTestHelpers.randomInterval;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
 
 @ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, numClientNodes = 4)
 public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
@@ -130,10 +135,6 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         assertAcked(client().execute(ILMActions.PUT, putLifecycleRequest).actionGet());
     }
 
-    @TestIssueLogging(
-        value = "org.elasticsearch.cluster.service.MasterService:TRACE",
-        issueUrl = "https://github.com/elastic/elasticsearch/issues/136585"
-    )
     public void testILMDownsampleRollingRestart() throws Exception {
         final InternalTestCluster cluster = internalCluster();
         cluster.startMasterOnlyNodes(1);
@@ -172,6 +173,18 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         startDownsampleTaskViaIlm(sourceIndex, targetIndex);
         assertBusy(() -> assertTargetIndex(cluster, targetIndex, indexedDocs));
         ensureGreen(targetIndex);
+        // We assert that ILM to successfully completed the phase
+        assertBusy(() -> {
+            ExplainLifecycleResponse explainLifecycleResponse = safeGet(
+                client().execute(
+                    ExplainLifecycleAction.INSTANCE,
+                    new ExplainLifecycleRequest(TimeValue.THIRTY_SECONDS).indices(targetIndex)
+                )
+            );
+            IndexLifecycleExplainResponse ilmExplain = explainLifecycleResponse.getIndexResponses().get(targetIndex);
+            assertThat(ilmExplain, notNullValue());
+            assertThat(ilmExplain.getStep(), equalTo(PhaseCompleteStep.NAME));
+        }, 60, TimeUnit.SECONDS);
     }
 
     private void startDownsampleTaskViaIlm(String sourceIndex, String targetIndex) throws Exception {

--- a/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
+++ b/x-pack/plugin/downsample/src/internalClusterTest/java/org/elasticsearch/xpack/downsample/ILMDownsampleDisruptionIT.java
@@ -14,6 +14,7 @@ import org.elasticsearch.action.admin.indices.get.GetIndexResponse;
 import org.elasticsearch.action.admin.indices.settings.get.GetSettingsRequest;
 import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest;
 import org.elasticsearch.action.downsample.DownsampleConfig;
+import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.core.TimeValue;
@@ -23,6 +24,7 @@ import org.elasticsearch.index.mapper.DateFieldMapper;
 import org.elasticsearch.index.query.MatchAllQueryBuilder;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
+import org.elasticsearch.test.ClusterServiceUtils;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.InternalTestCluster;
 import org.elasticsearch.xcontent.XContentBuilder;
@@ -30,14 +32,10 @@ import org.elasticsearch.xcontent.XContentFactory;
 import org.elasticsearch.xpack.aggregatemetric.AggregateMetricMapperPlugin;
 import org.elasticsearch.xpack.ccr.Ccr;
 import org.elasticsearch.xpack.core.LocalStateCompositeXPackPlugin;
-import org.elasticsearch.xpack.core.ilm.ExplainLifecycleRequest;
-import org.elasticsearch.xpack.core.ilm.ExplainLifecycleResponse;
-import org.elasticsearch.xpack.core.ilm.IndexLifecycleExplainResponse;
 import org.elasticsearch.xpack.core.ilm.LifecyclePolicy;
 import org.elasticsearch.xpack.core.ilm.LifecycleSettings;
 import org.elasticsearch.xpack.core.ilm.Phase;
 import org.elasticsearch.xpack.core.ilm.PhaseCompleteStep;
-import org.elasticsearch.xpack.core.ilm.action.ExplainLifecycleAction;
 import org.elasticsearch.xpack.core.ilm.action.ILMActions;
 import org.elasticsearch.xpack.core.ilm.action.PutLifecycleRequest;
 import org.elasticsearch.xpack.ilm.IndexLifecycle;
@@ -52,7 +50,9 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import static org.elasticsearch.core.Strings.format;
@@ -173,18 +173,14 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
         startDownsampleTaskViaIlm(sourceIndex, targetIndex);
         assertBusy(() -> assertTargetIndex(cluster, targetIndex, indexedDocs));
         ensureGreen(targetIndex);
-        // We assert that ILM to successfully completed the phase
-        assertBusy(() -> {
-            ExplainLifecycleResponse explainLifecycleResponse = safeGet(
-                client().execute(
-                    ExplainLifecycleAction.INSTANCE,
-                    new ExplainLifecycleRequest(TimeValue.THIRTY_SECONDS).indices(targetIndex)
-                )
-            );
-            IndexLifecycleExplainResponse ilmExplain = explainLifecycleResponse.getIndexResponses().get(targetIndex);
-            assertThat(ilmExplain, notNullValue());
-            assertThat(ilmExplain.getStep(), equalTo(PhaseCompleteStep.NAME));
-        }, 60, TimeUnit.SECONDS);
+        // We assert that ILM successfully completed the phase
+        logger.info("Waiting for ILM to complete the phase for index [{}]", targetIndex);
+        ClusterServiceUtils.addTemporaryStateListener(clusterState -> {
+            IndexMetadata indexMetadata = getIndexMetadataByIndexName(clusterState, targetIndex);
+            // We expect a single project available in this cluster
+            return indexMetadata.getLifecycleExecutionState() != null
+                && Objects.equals(indexMetadata.getLifecycleExecutionState().step(), PhaseCompleteStep.NAME);
+        });
     }
 
     private void startDownsampleTaskViaIlm(String sourceIndex, String targetIndex) throws Exception {
@@ -232,5 +228,20 @@ public class ILMDownsampleDisruptionIT extends DownsamplingIntegTestCase {
                 assertTrue(targetIndexSearch.getHits().getHits().length > 0);
             }
         );
+    }
+
+    /**
+     * We assume the index exists and its index name is unique.
+     * @return the index metadata found in any project that contains an index with this name.
+     */
+    private IndexMetadata getIndexMetadataByIndexName(ClusterState clusterState, String indexName) {
+        AtomicReference<IndexMetadata> indexMetadata = new AtomicReference<>();
+        clusterState.forEachProject(projectState -> {
+            if (projectState.metadata().hasIndex(indexName)) {
+                indexMetadata.set(projectState.metadata().index(indexName));
+            }
+        });
+        assertThat(indexMetadata.get(), notNullValue());
+        return indexMetadata.get();
     }
 }

